### PR TITLE
[air] Use filesystem wrapper to exclude files from upload

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Ray Tune uploads experiment state using pyarrow. When cloud checkpointing is configured, the driver will exclude any trial-level checkpoints. Pyarrow does not natively support file exclusion, though - instead, we repeatedly call `pyarrow.fs.copy_files` on single non-excluded files.

This seems to be inefficient as the connection to the remote filesystem is opened and closed repeatedly. It also means we can never leverage multi-threaded upload. This PR implements a custom fsspec-based local filesystem that excludes files on the selector level. Thus, we can call pyarrow.fs.copy_files exactly once, with a selector that does not see the excluded files. 

Edit: [See here for benchmark results](https://github.com/ray-project/ray/pull/34102#issuecomment-1517846074)

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by krfricke** [comment]
I'll run the restore example with these wheels to check if it improves performance

**Comment by krfricke** [comment]
It does not seem to significantly impact the upload time. I still think this is generally a cleaner version to go for, but I'll look more into this.
Converting to draft for now pending further investigation.

**Comment by krfricke** [comment]
I ran a few benchmarks to compare different versions of our code.

*Data*: 5 files a 2MB in one folder

Compared methods:

- `awscli_sync`: Direct usage of `aws s3 sync`
- `pyarrow_threads`: Direct usage of `pyarrow.fs.copy_files(source, target, use_threads=False)`
- `pyarrow_no_threads`: Direct usage of `pyarrow.fs.copy_files(source, target, use_threads=True)`
- `air_default`: Usage of AIR's `upload_to_uri(source, target)` utility. Defaults to no-threads for S3
- `air_exclude_threads`: `upload_to_uri(source, target, exclude=something)` and with threading _enabled_
- `air_exclude_no_threads`: `upload_to_uri(source, target, exclude=something)` and with threading disabled (default)
- `air_patched_threads`: Version of this PR with threading enabled
- `air_patched_no_threads`: Version of this PR with threading disabled.

Results:

```
Results:
    awscli_sync:         0.76 (0.01) seconds
    pyarrow_threads:     0.69 (0.04) seconds
    pyarrow_no_threads:  1.19 (0.07) seconds
    air_default:         1.19 (0.08) seconds
    air_exclude_threads:         1.43 (0.02) seconds
    air_exclude_no_threads:      1.32 (0.03) seconds
    air_patched_threads:         0.53 (0.01) seconds
    air_patched_no_threads:      1.23 (0.08) seconds
```

The TLDR of these results is that threading gives an (obvious) advantage.

However, due to https://github.com/apache/arrow/issues/32372, we currently can't enable threading per default.

But, assuming https://github.com/apache/arrow/issues/32372 is fixed at some point, our current `exclude` logic can't leverage threading, as we call `pyarrow.fs.copy_files` multiple times, sequentially. Thus we see that `air_exclude_threads` has the same bad performance as `pyarrow_no_threads` or `air_exclude_no_threads`

With the patch from this PR, we can regain the threading benefit, once the pyarrow issue is resolved, as shown in the good performance of `air_patched_threads`.

Thus, we should aim to land this PR.

I'll work on a workaround for the pyarrow issue separately.

File: python/ray/tune/execution/experiment_state.py [hunk 1]
```diff
@@ -266,17 +272,41 @@ def sync_up(self, force: bool = False, wait: bool = False) -> bool:
...
+                import fsspec
+            except Exception:
+                fsspec = None
+
+            fsspec_msg = ""
+            if fsspec is None:
+                fsspec_msg = (
+                    "If your data is small, try installing fsspec "
+                    "(`pip install fsspec`) for more efficient local file parsing. "
+                )
```
Comment (justinvyu) [inline python/ray/tune/execution/experiment_state.py:293]: Should `ffspec` always be recommended, or is it only beneficial for many, small files? Does it perform worse for few, large files?

File: python/ray/tune/execution/experiment_state.py [hunk 2]
```diff
@@ -138,6 +137,13 @@ def __init__(
         # Upload triggered by trial checkpoints
         self._sync_every_n_trial_checkpoints = sync_every_n_trial_checkpoints
         self._trial_num_checkpoints_since_last_sync: Dict[Trial, int] = Counter()
+
+        self._slow_sync_threshold = float(
+            os.environ.get(
+                "TUNE_WARN_SLOW_EXPERIMENT_CHECKPOINT_SYNC_THRESHOLD_S", "30"
```
Comment (justinvyu) [inline python/ray/tune/execution/experiment_state.py:143]: I am wondering if we should always warn users if they don't have `fsspec`, rather than just if their exp syncing is taking a while? The threshold might mask the problems and the user would never know to do something.

Will there be a separate warning checking if using any cloud storage + no `fsspec`?

**Review by justinvyu — COMMENTED** [review]
Looks good! Could we also add a unit test that tests `upload_to_uri` with and without `fsspec`? Maybe you can monkeypatch fsspec to `None` to switch to default.

File: python/ray/tune/execution/experiment_state.py (hunk 2 above)
Comment (krfricke) [reply python/ray/tune/execution/experiment_state.py:143]: Yes! I have added the warnings in the follow-up PR: https://github.com/ray-project/ray/pull/34663

File: python/ray/tune/execution/experiment_state.py (hunk 1 above)
Comment (krfricke) [reply python/ray/tune/execution/experiment_state.py:304]: Initial benchmarks suggested that it always performed better. After https://github.com/ray-project/ray/pull/34663 any implementation efficiencies will be hugely shadowed by being able to leverage multi threaded uploads, so we should always recommend fsspec (we'll be printing warnings in https://github.com/ray-project/ray/pull/34663 if it's not installed).

**Comment by krfricke** [comment]
Added a test, ptal!

**Review by justinvyu — APPROVED** [review]
Thanks for clarifying! One nit on another test case. Also, what's the main difference between the 2 unit tests that were added?

File: python/ray/air/tests/test_remote_storage.py [hunk 3]
```diff
@@ -131,6 +131,46 @@ def test_upload_exclude_multimatch(temp_data_dirs):
...
 
+@pytest.mark.parametrize("no_fsspec", [False, True])
+def test_upload_local_exclude_multi(temp_data_dirs, no_fsspec):
+    if no_fsspec:
+        with patch("ray.air._internal.remote_storage.fsspec", None):
+            return test_upload_local_exclude_multi(temp_data_dirs, no_fsspec=False)
+
+    tmp_source, tmp_target = temp_data_dirs
+
+    upload_to_uri(tmp_source, tmp_target, exclude=["*_exclude.txt", "*_exclude/*"])
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/air/tests/test_remote_storage.py:142]: The tests were copied from existing tests that test the same behavior with an in-memory filesystem. The main difference is that in the second case, multiple patterns are tested (not just one).

The `exclude/` pattern won't work as it will not match the files contained within the directory - it always has to be `exclude/*`

**Review by maxpumperla — APPROVED** [review]
